core/src/main/java/com/linecorp/armeria/common/HttpRequestBuilder.java

```diff
@@ -0,0 +1,397 @@
+        return this;
+    }
+
+    /**
+     * Sets multiple query params for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/endpoint")
+     *            .queryParams(QueryParams.of("from", "foo", "limit", 10))
+     *            .build(); // GET `/endpoint?from=foo&limit=10`
+     * }</pre>
+     * @see QueryParams
+     */
+    public HttpRequestBuilder queryParams(QueryParams queryParams) {
+        requireNonNull(queryParams, "queryParams");
+        queryParams.forEach((BiConsumer<String, String>) queryParamsBuilder::set);
+        return this;
+    }
+
+    /**
+     * Sets a cookie for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/")
+     *            .cookie(Cookie.of("cookie", "foo"))
+     *            .build();
+     * }</pre>
+     * @see Cookie
+     */
+    public HttpRequestBuilder cookie(Cookie cookie) {
+        cookies.add(requireNonNull(cookie, "cookie"));
+        return this;
+    }
+
+    /**
+     * Sets multiple cookies for this request. For example:
+     * <pre>{@code
+     * HttpRequest.builder()
+     *            .get("/")
+     *            .cookies(Cookies.of(Cookie.of("cookie1", "foo"),
+     *                                Cookie.of("cookie2", "bar")))
+     *            .build();
+     * }</pre>
+     * @see Cookies
+     */
+    public HttpRequestBuilder cookies(Cookies cookies) {
+        this.cookies.addAll(requireNonNull(cookies, "cookies"));
+        return this;
+    }
+
+    /**
+     * Creates a new {@link HttpRequest}.
+     */
+    public HttpRequest build() {
+        final RequestHeaders requestHeaders = requestHeaders();
+        if (content == null || content.length() == 0) {
+            if (content != null) {
+                content.close();
+            }
+            if (httpTrailersBuilder.isEmpty()) {
+                return new EmptyFixedHttpRequest(requestHeaders);
+            }
+            return new OneElementFixedHttpRequest(requestHeaders, httpTrailersBuilder.build());
+        }
+        if (httpTrailersBuilder.isEmpty()) {
+            return new OneElementFixedHttpRequest(requestHeaders, content);
+        }
+        return new TwoElementFixedHttpRequest(requestHeaders, content, httpTrailersBuilder.build());
+    }
+
+    private RequestHeaders requestHeaders() {
+        requestHeadersBuilder.path(fullPath());
+        if (!cookies.isEmpty()) {
+            requestHeadersBuilder.set(COOKIE, Cookie.toCookieHeader(cookies));
+        }
+        if (content == null || content.length() == 0) {
+            requestHeadersBuilder.remove(CONTENT_LENGTH);
+        } else {
+            requestHeadersBuilder.setInt(CONTENT_LENGTH, content.length());
+        }
+        return requestHeadersBuilder.build();
+    }
+
+    private String fullPath() {
+        final StringBuilder pathBuilder = new StringBuilder(requireNonNull(this.path, "path"));
```

Comment:
can remove `this.` and how about raising an `IllegalStateException` rather than NPE?
```java
checkState(path != null, "path must be set.")
```